### PR TITLE
Database Migrations

### PR DESCRIPTION
**Pull request by Cyberboss**
Long time coming

Configs:
 - `MIGRATIONS` (flag): To use migration system, schema will automatically try to upgrade to highest code version at SS init
 - `SINGLE_MIGRATION_TRANSACTION` (flag): If enabled, migrations are applied as one single transaction as opposed to one transaction per migration

Currently there's no way to migrate downwards (although the code exists and we should enforce people write the queries for them), not sure what our use case is for that at this point

See the README.md in modules/database_migration and my WIP work on schema version 4.7 as docs/examples

### CODE REVIEW

**Comment by Cyberboss** [comment]
>Forgot my commits at work

**Comment by optimumtact** [comment]
are you actually making schema changes to the db? if so, could you list them and the reasonings.

**Comment by Cyberboss** [comment]
There are no schema changes made by this. This is a framework for running future changes automatically via game code

File: code/modules/database_migrations/4.7_LegacySchema.dm [hunk 1]
```diff
@@ -0,0 +1,340 @@
+/datum/database_migration/LegacySchema
+	schema_version_major = 4
+	schema_version_minor = 7
+
+/datum/database_migration/LegacySchema/Up()
+	//original v4.7 schema creation scripts here: https://github.com/tgstation/tgstation/tree/f28cd100991f4a35c8028b63aa1b40f3c2393cff/SQL
```
Comment (Cyberboss) [inline code/modules/database_migrations/4.7_LegacySchema.dm]: @Jordie0608, verify I'm not fucking anything up here please

**Comment by Cyberboss** [comment]
Travis lies about success

**Comment by Cyberboss** [comment]
No wait, the migrations ran fine, but some other thing failed

**Comment by optimumtact** [comment]
NANI

**Comment by Cyberboss** [comment]
I need to merge @Jordie0608's changes, but after that should be good

**Comment by MrStonedOne** [comment]
@Cyberboss 

Allow me to throw a wrench in your idea.

What about the other servers?

What happens if, say, a ban table change gets merged, and deployed to terry, triggering an automatic migration?

Now bagil and sybil can't check bans.

Allow me to throw another wrench in your idea:

On /tg/ at least, The mysql users `sybil` `basil` `terry` and `events` do not have any DDL permissions.

In fact they mostly only have insert and select permissions, with selective column level update permissions based on what columns game code actually has code to update, and they only have delete permissions on tables game code actually has code for that deletes rows.

Anywho this should still be super handy for downstreams, but i'm afraid I won't be able to enable it for here.

**Comment by Cyberboss** [comment]
I figured as much, this is mainly to help out local development though

**Comment by optimumtact** [comment]
@MrStonedOne it would make the upgrade process simpler for you though, since people can write the transition code for you.

We would still let you decide the time of merging but it should surely make your job easier.

**Comment by MrStonedOne** [comment]
I don't really trust people to write good migration code.

Mainly because they almost always require creating views to do properly.

Also our system for upgrades is kinda lame. I'd suggest having every modification to a table require changing that tables name, so you can properly do old views onto new tables while allowing new code to write to the new table. That or some system for appending _new to table names during migration periods. 

It's lame when I have to choose between breaking read and write for unmigrated servers or breaking write for all servers until all are migrated. If i could easily a/b the changed tables so the unmigrated servers try to access one (that I can make a view), and migrated servers try to access another (the real migrated table), I could do views without breaking write functionality on the migrated servers.


Food for thought.

**Pull request closed without merging**